presto-main/src/main/java/com/facebook/presto/execution/SqlStageExecution.java

```diff
@@ -483,8 +519,21 @@ public synchronized Set<RemoteTask> scheduleSplits(InternalNode node, Multimap<P
         }
         else {
             task = tasks.iterator().next();
-            task.addSplits(splits);
+            boolean nodeNotShutDown = task.addSplits(splits);
+
+            if (isEnableGracefulShutdown || isRetryOfFailedSplitsEnabled) {
+                if (!nodeNotShutDown) {
+                    Optional<InternalNode> firstActiveNode = nodes.stream().filter(n -> n.getNodeIdentifier() != node.getNodeIdentifier()).findFirst();
+                    if (!firstActiveNode.isPresent()) {
+                        throw new RuntimeException(String.format("Unable to find the next active nodes to assign the retried splits"));
+                    }
+                    Collection<RemoteTask> tasksToAssignRetriedSplits = this.tasks.get(firstActiveNode.get());
```

Comment:
Can tasksToAssignRetriedSplits.size() > 1 ? Even within the same stage -- I imagined that each node has exactly one task for a stage so for the new candidate node "firstActiveNode" -- shouldn't there be exactly one task ? (Actually are we guaranteed to find a task ?) Indeed, you pick the first task from this collection on the next line.